src/stacked_borrows.rs

```diff
@@ -565,8 +613,8 @@ impl Stacks {
             range.size.bytes()
         );
         let global = &*extra.borrow();
-        self.for_each(range, move |offset, stack| {
-            stack.access(AccessKind::Read, tag, Pointer::new(alloc_id, offset), global)
+        self.for_each(range, move |_offset, stack| {
+            stack.access(AccessKind::Read, tag, (alloc_id, range), global)
```

Comment:
This loses precision -- `range` will be the range of the *entire* access, but `access` here is called for just a particular sub-range starting at `offset`. So the resulting error might even be factually wrong, by stating a too large range.

Reply:
The memory access that we're handling is possibly granted by `tag` over the entirety of `range` and of the same kind throughout. It doesn't matter where in that range the tag stops granting access, what I'm suggesting reporting is that `tag` doesn't grant the permission over all of `range`.

Perhaps it would be _more_ precise to state exactly what range the tag doesn't grant sufficient permission for, but I don't see how the diagnostics this reports are wrong.

Reply:
Fair -- I guess it depends on how one interprets "tag X is invalid for access at start..end", i.e. whether there is an implicit universal or existential quantification.

But then the question remains, why regress the currently more precise error messages? If you want to report a range, we can make `for_each` provide the precise range.

Reply:
Can you come up with some code where this improved precision changes the diagnostic? I feel like this has something to do with freezing, but I can't come up with an example at the moment.

Reply:
No, freezing won't trigger this. You need to get the underlying `RangeMap` to split the ranges because the stacks diverge. Roughly like:
```rust
let mut pair = (0i32, 1i32);
let ptr = &mut pair as *mut _;
pair.0 += 1;
let _val = *ptr; // now just the range 0..4 of locations is "bad"
```

Reply:
I'm still confused. I modified the program to this, to provoke an error:
```rust
fn main() {
    let mut pair = (0i32, 1i32);
    let ptr = &mut pair as *mut _;
    let another = &mut pair.1;
    unsafe {
        *ptr = (1i32, 1i32);
    }
}
```
But the diagnostic comes out as:
```
 --> src/main.rs:6:9
  |
6 |         *ptr = (1i32, 1i32);
  |         ^^^^^^^^^^^^^^^^^^^ attempting a write access of alloc953[0x4..0x8] via tag <untagged>, but <untagged> does not exist in the borrow stack for this memory range
  |
```
That seems correct to me.

If I modify the program a bit to be something I'm sure will write across a range which is split:
```rust
fn main() {
    let mut pair = [0i32, 1i32];
    let ptr = &mut pair as *mut i32 as *mut i64;
    let another = &mut pair[1];
    unsafe {
        *ptr = 0i64;
    }
}
```
Then I get this:
```
 --> src/main.rs:6:9
  |
6 |         *ptr = 0i64;
  |         ^^^^^^^^^^^ attempting a write access of alloc953[0x0..0x8] via tag <untagged>, but <untagged> does not exist in the borrow stack for this memory range
  |
```
It sounds to me like you want that to say `alloc953[0x4..0x8]`, but I think a user is likely to assume that's a Miri bug. This is clearly a write of size 8. Why is Miri telling me about a write of size 4?

Reply:
I need to sleep, but questions like this make me want to draw diagrams of the allocation in question. Maybe that's a good thing to do eventually....

Reply:
> `*ptr = (1i32, 1i32)`

this gets compiled to two separate assignments:
```
((*_2).0: i32) = const 1_i32;    // scope 4 at [src/main.rs:6:9: 6:28]
((*_2).1: i32) = const 1_i32;    // scope 4 at [src/main.rs:6:9: 6:28]
```
That's why you are not seeing the range you are expecting.
Use something like this instead:
```rust
let newval = (1i32, 1i32);
*ptr = newval
```

> I modified the program to this, to provoke an error:

FWIW my example [also provokes an error](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e092a4527770a4175ebefce547263d82)?

> It sounds to me like you want that to say alloc953[0x4..0x8], but I think a user is likely to assume that's a Miri bug. This is clearly a write of size 8. Why is Miri telling me about a write of size 4?

Aliasing is strictly per-location, so an 8-byte access are simply 8 independent 1-byte accesses. It is for pure performance reasons that Miri groups neighboring borrow stacks together and often handles entire many-byte accesses via a single borrow stack operation.

Reply:
My concern is printing a span which _clearly_ points to an access of size 8, then saying something about a smaller range. I think the best approach is to print the exact range in the error, and if the range was split, include a note which explains to the user that this error is just for the first part of the whole operation that Miri has detected is invalid. Does that sound good to you?

Reply:
That would be a possibility.

However I increasingly think we are going down the wrong path here. Stacked Borrows should be thought of as having an independent borrow stack for each memory location. The fact that Miri fuses neighboring stacks is an optimization and should not affect user-visible behavior. As such, a Stacked Borrows violation always occurs at a *single* memory address, never at a range -- and IMO that is how we should report it.

Reply:
What about this:
```
  |
6 |         *ptr = 0;
  |         ^^^^^^^^ attempting a write access at alloc953+0x4 via tag <1234>, but <1234> does not exist in the borrow stack for this address

note: this error was detected while handling a write to alloc953[0x0..0x8]
```
The error is exactly correct, and in combination with the note, the user can reasonably deduce what is going on here without fishing through their source code to figure out the size of the pointed-to type.

Reply:
I assume you are concerned that putting the entire write range into the main message together with the exact address will make that message too long?

We don't have such notes in other errors, but also most errors don't contain nearly as much information as Stacked Borrows errors, so this is fine for me. It might be tricky to implement though.^^

The fact that it talks about a "write" twice seems somewhat redundant, not sure if that is needed.

Reply:
Yeah, I'm concerned about the size of the errors. I have yet more ideas on things to add, and already they are big enough that adding to the one long line is becoming an issue.

I implemented two versions of this. One just sticks another line into the existing `helps` machinery, but that generates a whole extra span. The latest commit moves the extra line into the original span, with `span_label` and looks like this:
```
error: Undefined Behavior: trying to reborrow alloc953+0x0 with Unique permission, but parent tag <1981> only grants SharedReadOnly permission for this memory
 --> src/main.rs:5:18
  |
5 |         let _z = &mut *y;
  |                  ^^^^^^^
  |                  |
  |                  trying to reborrow alloc953+0x0 with Unique permission, but parent tag <1981> only grants SharedReadOnly permission for this memory
  |                  this is a reborrow of alloc953[0x0..0x4]
  |
```

I prefer this one, but it's hard to tell if I'm abusing the error reporting machinery. Though I'm not sure how much of a leg the existing code has to stand on, seeing that it pastes the whole error message into the title ;)

Reply:
No strong opinion on my side. However I think I'd prefer wording like "this is part of a reborrow at alloc953[0x0..0x4]" or so.

It also seems a bit odd to sometimes use `[...]` and sometimes `+` to join the index to the allocation ID?

Reply:
Pronouns are hard. I wanted "this" to refer to the span, but I'm going to go with "this error occurs as part of a..." so there's no ambiguity.

You're right. I simply did not think of putting one value in the brackets.